src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py

```diff
@@ -60,77 +68,42 @@ def __call__(self, t):
         :obj:`RandomVariable`
             Estimate of the states at time ``t``.
         """
+
+        # Recursive evaluation (t can now be any array, not just length 1)
         if not np.isscalar(t):
-            # recursive evaluation (t can now be any array, not just length 1!)
             return _RandomVariableList([self.__call__(t_pt) for t_pt in np.asarray(t)])
 
+        # t is left of our grid -- raise error
+        # (this functionality is not supported yet)
         if t < self.locations[0]:
             raise ValueError(
                 "Invalid location; Can not compute posterior for a location earlier "
                 "than the initial location"
             )
 
+        # Early exit if t is in our grid -- no need to interpolate
         if t in self.locations:
-            idx = (self.locations <= t).sum() - 1
+            idx = self._find_previous_or_same_index(t)
             discrete_estimate = self.state_rvs[idx]
             return discrete_estimate
 
-        if self.locations[0] < t < self.locations[-1]:
-            pred_rv = self._predict_to_loc(t)
-            if self._with_smoothing:
-                smoothed_rv = self._smooth_prediction(pred_rv, t)
-                return smoothed_rv
-            else:
-                return pred_rv
-
-        # else: t > self.locations[-1], which case we just predict.
-        return self._predict_to_loc(t)
-
-    def _predict_to_loc(self, loc):
-        """Predict states at location `loc` from the closest, previous state."""
-        previous_idx = (self.locations < loc).sum() - 1
-        previous_loc = self.locations[previous_idx]
-        previous_rv = self.state_rvs[previous_idx]
-
-        pred_rv, _ = self.gauss_filter.predict(
-            rv=previous_rv, start=previous_loc, stop=loc
-        )
-        return pred_rv
+        return self.interpolate(t)
 
-    def _smooth_prediction(self, pred_rv, loc):
-        """Smooth the predicted state at location `loc` using the next closest."""
-        next_idx = (self.locations < loc).sum()
-        next_loc = self.locations[next_idx]
-        next_rv = self._state_rvs[next_idx]
-
-        # Intermediate prediction
-        predicted_future_rv, info = self.gauss_filter.predict(
-            rv=pred_rv,
-            start=loc,
-            stop=next_loc,
-        )
-        crosscov = info["crosscov"]
-
-        curr_rv, _ = self.gauss_filter.smooth_step(
-            pred_rv, predicted_future_rv, next_rv, crosscov
-        )
-        return curr_rv
-
-    def __len__(self):
-        return len(self.locations)
-
-    def __getitem__(self, idx):
-        return self.state_rvs[idx]
+    @abc.abstractmethod
+    def interpolate(self, t):
+        """Evaluate the posterior at a measurement-free point."""
+        raise NotImplementedError
```

Comment:
this is one of two differences in filtering and smoothing posterior -- dense evaluation (interpolation).